torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+
+                for rank, params in enumerate(param_lists):
+                    param_group_rank = copy.copy(param_group)
+                    param_group_rank["params"] = params
+                    self._partition_parameters[rank].append(param_group_rank)
+
+        return self._partition_parameters
+
+    @property
+    def per_device_params(self) -> Dict[torch.device, List[List[Parameter]]]:
+        """Sorted list of all the params, first per device then per rank.
+
+        Within a list params are sorted per number of elements to allow for an easy bucketing.
+        """
+        if len(self._per_device_params) == 0:
+            # Go through all params, log them per device
+            # The ordering is important here, needs to be the same on all ranks
+            # So that ulterior broadcast calls are matching
+            for param_group in self.param_groups:
+                for param in param_group["params"]:
+                    device = param.device
+                    if self._per_device_params.get(device) is None:
+                        self._per_device_params[device] = [[] for _ in range(self.world_size)]
+                    self._per_device_params[device][self.param_to_rank[param]] += [param]
+
+            # Sort param_lists by size
+            for k in self._per_device_params.keys():
+                for r in self._per_device_params[k]:
+                    r.sort(key=lambda x: x.numel())
+
+        return self._per_device_params
+
+    @property
+    def param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """Look up table to match a given param with a data parallel rank"""
+        if len(self._param_rank) == 0:
+            for rank, param_groups in enumerate(self.partition_parameters()):
+                for param_group in param_groups:
+                    for param in param_group["params"]:
+                        self._param_rank[param] = rank
+        return self._param_rank
+
+    def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) -> Optional[float]:
+        """Performs a single optimization step (parameter update).
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model and
+                returns the loss. Optional for most optimizers.
+        Returns:
+            optional loss, depends on the underlying optimizer
+
+        .. note: Any extra parameter is passed to the base optimizer as-is"""
+
+        # Sync oss param_groups attributes in case they've been updated by a scheduler.
+        self._sync_param_groups()
+
+        # Run the optimizer step on this shard only:
+        if closure is not None:
+            loss = self.optim.step(closure=closure, **kwargs)  # type: ignore
+        else:
+            loss = self.optim.step(**kwargs)
+
+        # Free the other grads
+        # ideally this would not be needed, handled in the reduce step by a zero-matching DDP engine
+        self._free_other_grads()
```

Comment:
I removed that upstream so probably better to remove it here also, but the DDP that we use with this optimizer does remove the gradients on the fly, because that's basically what gradient sharding means. In essence sharding implies that each rank does not have access to everything, it could be made a little more seamless with rpc/remote references if we want to keep this usecase, but without that it's pretty orthogonal to the end goal

Reply:
Once we do implement gradient sharding, this makes sense. However, the user can only use this optimizer standalone with PyTorch DDP only for optimizer sharding. In that case, we're deviating from the contract we have for other optimizers where gradients have to be manually cleared by the user.

Reply:
I've removed it from here, it was also removed from fairscale some time ago, this PR is a bit too old. There's no contract though AFAIK, this ZeroRedundancyOptimizer+ShardedDDP means that most of the grads are wiped, that's by design or there's not much to do as far as the name goes ("ZeroRedundancy" means, well, that some things will be removed)

Reply:
I just realized that it may not have been clear, but fairscale is doing gradient sharding already, ZeroRedundancyOptimizer is one piece of the puzzle. In any case this particular line is gone, so I think we agree :)